Tensile/AsmRegisterPool.py

```diff
@@ -181,22 +182,50 @@ def isRangeAvailable(self, start, size, preventOverflow=-1, wantedStatus=Status.
 
     return True
 
-  def findFreeRange(self, size, alignment, preventOverflow=-1, wantedStatus=Status.Available):
+
+  def findFreeRange(
+    self, 
+    size: int, 
+    alignment: int, 
+    preventOverflow: int=-1, 
+    wantedStatus: Status=Status.Available
+  ) -> Optional[int]:
+    """Find a free range of registers of a given size, alignment, and status from the register pool.
+
+    Args:
+        size: The required size of the register block.
+        alignment: The alignment requirement for the block. Each block must start at an index
+            that is a multiple of this value.
+        preventOverflow: Flag to prevent overflow. If set to -1, it uses the default value.
+        wantedStatus: The status of the block to be found.
+
+    Returns:
+        1. The starting index of the block if found. 
+        2. None if preventOverflow is set and no block is found.
+        3. The start of the last available block rounded up to the alignment if preventOverflow is
+            not set and no block is found. 
+    """
     if preventOverflow == -1:
       preventOverflow = self.defaultPreventOverflow
 
-    for i in range(len(self.pool)+1):
-      if i % alignment != 0:
-        continue
-      if self.isRangeAvailable(i, size, preventOverflow, wantedStatus):
-        return i
+    left, right = 0, 0
+    while right < len(self.pool):
```

Comment:
What is the main cause of bad performance of the original code?
i % alignment?
calling isRangeAvailable?

So far, your code looks like not handling preventOverflow.
If i%alignment is the primary cause, we can just call isRangeAvailable here.

Reply:
Calling `isRangeAvailable` is the cause of bad performance. In its current form, it uses nested for loops to achieve the same outcome as the solution presented here, which uses a single pass through the register pool (`self.pool`).

I do believe we are appropriately handling preventOverflow because the loop will exit of the right index exceeds the pool size.

Reply:
Difference is this part

    if preventOverflow and end > len(self.pool):
      return False

    end = min(end, len(self.pool))

if preventOverflow == false, isRangeAvailable does not check start + size.
It returns true as long as from start to end = min(end, len(self.pool)) is available.

We can add extra if condition to follow this logic.
(however, it does not make much sense we do not check size....)

Reply:
isRangeAvailable will continue to check start+size if preventOverflow==False, so long as start+size isn't larger than the pool. In the case that `end <= len(self.pool)`, we loop over every entry from start to end, always starting from the beginning of an alignment.

In the code update, we're effectively collecting the alignment step in the same iteration where we search for the range. If at anytime the block size matches the requested size, we return the left index. 

Either way, I will write some tests to ensure the behaviour is the same as before.

Reply:
I am OK with your change, but what I am saying is isRangeAvailable seems incorrect in that case.